src/backend/cdb/cdbdtxrecovery.c

```diff
@@ -153,6 +155,18 @@ recoverTM(void)
 	elogif(Debug_print_full_dtm, LOG,
 		   "Starting to Recover DTM...");
 
+	/*
+	 * We'd better terminate residual QE processes to avoid potential issues,
+	 * e.g. shared snapshot collision, etc. We do soft-terminate here so it is
+	 * still possible there are residual QE processes but it's better than doing
+	 * nothing.
+	 *
+	 * We just do this when there was abnormal shutdown on master or standby
+	 * promote, else mostly there should not have residual QE processes.
+	 */
+	if (*shmCleanupBackends)
+		TerminateMppBackends();
```

Comment:
Will shmCleanupBackends become false only at shared memory reset without relaunching postmaster?

Reply:
> Is the situation possible, when the bg worker is relaunched but without master and shared memory reset? In this case `shmCleanupBackends ` will remain true.

No, the variable is set to false when the shared memory is initialized outside of the postmaster.

Reply:
> Will shmCleanupBackends become false only at shared memory reset without relaunching postmaster?

Not only, but also when initializing shared memory outside of the postmaster.

Reply:
Despite the shmem initialization, when `shmCleanupBackends` is set to false? Or it's ok when it's mostly always true?

Reply:
> Despite the shmem initialization, when `shmCleanupBackends` is set to false? Or it's ok when it's mostly always true?

Yes, if we want to kill orphaned backends on executors not only when the coordinator crashes, but also when it restarts.

Reply:
> Yes, if we want to kill orphaned backends on executors not only when the coordinator crashes, but also when it restarts.

In other words we may call TerminateMppBackends unconditionally? Because using [this](https://github.com/arenadata/gpdb/blob/e655547faf9ff63aa4184bd9b344ffb6347ac090/src/backend/cdb/cdbdtxrecovery.c#L772) flag is sufficient for that purposes. Or not?

Reply:
> we may call TerminateMppBackends unconditionally?

yes

Reply:
@bimboterminator1, I think your question makes sense.

@RekGRpth, dtx process may catch `elog(ERROR` and will be restarted. Do we really want to kill all backends on QEs in this case?

Reply:
If dtx recovery process  is relaunched without shared memory reset the `shmCleanupBackends` flag will remain true. However, if dtx process failure takes place after `shmDtmStarted` has been set to true, the `shmDtmStarted` will also remain true, but the TerminateMppBackends won't be called because of latter [flag](https://github.com/arenadata/gpdb/blob/e655547faf9ff63aa4184bd9b344ffb6347ac090/src/backend/cdb/cdbdtxrecovery.c#L772).

Reply:
Thank you for explanation!